### PR TITLE
Update Gradient namespace

### PR DESCRIPTION
**Pull request by gautierronan**
Following the benchmark of `dx.RecursiveCheckpointAdjoint` (see below), this PR removes the adjoint method altogether (at least for now).

There are now two available gradient methods:
 - `Autograd` which correponds to the jax autograd, available in diffrax using `dx.DirectAdjoint`
 - `CheckpointAutograd` corresponding to `dx.RecursiveCheckpointAdjoint` (not available for our custom solvers, for now)

The first one seems fastest if memory is not an issue. The second one solves the memory problem similarly to the adjoint method, but probably better all-around.

NB: Also fix the `dq.expect` bug by using `jnp.einsum` and the `jnp.diff(..., prepend=...)` bug.

<details>
  <summary>Benchmark of gradients in diffrax on a small problem</summary>

```python
import jax
import jax.numpy as jnp
import diffrax as dx
import dynamiqs as dq
import functools
import timeit
import warnings


# create a simple quantum problem on which to test the adjoints
@functools.partial(jax.jit, static_argnums=(1, 2))
def run_mesolve(delta, adjoint, N):
    # make quantum problem
    a = dq.destroy(N)
    i = dq.eye(N)
    H = dq.dag(a) @ dq.dag(a) @ a @ a
    L = 0.1 * a @ a
    tsave = jnp.linspace(0.0, 10.0, 100)
    rho0 = dq.coherent_dm(N, 2.0)

    # initialize diffrax solver
    def vector_field(t, y, args):
        out = -1j * args[0] * H @ y + 0.5 * L @ y @ dq.dag(L) - 0.5 * dq.dag(L) @ L @ y
        return out + dq.dag(out)

    term = dx.ODETerm(vector_field)
    solver = dx.Dopri5()
    saveat = dx.SaveAt(ts=tsave)
    stepsize_controller = dx.PIDController(rtol=1e-5, atol=1e-7)

    # run solver
    with warnings.catch_warnings():
        warnings.simplefilter('ignore', UserWarning)
        solution = dx.diffeqsolve(
            term,
            solver,
            t0=tsave[0],
            t1=tsave[-1],
            dt0=None,
            y0=rho0,
            saveat=saveat,
            stepsize_controller=stepsize_controller,
            adjoint=adjoint,
            args=(delta,),
        )

    return dq.expect(a, solution.ys[-1]).real


# initialize parameters and grad function
N = 20
delta = 0.2
grad_mesolve = jax.grad(run_mesolve, argnums=0)

# time mesolve with BacksolveAdjoint (uncomment to test)
# print("BacksolveAdjoint")
# adjoint = dx.BacksolveAdjoint()
# %timeit grad_mesolve(delta, adjoint, N)

# time mesolve with DirectAdjoint (uncomment to test)
print("DirectAdjoint")
adjoint = dx.DirectAdjoint()
grad_mesolve(delta, adjoint, N) # jit compilation
%timeit -n 1 -r 1 grad_mesolve(delta, adjoint, N)

# time mesolve with RecursiveCheckpointAdjoint (uncomment to test)
print("RecursiveCheckpointAdjoint")
for n in range(10, 0, -1):
    print(f"== n = {n} ==")
    adjoint = dx.RecursiveCheckpointAdjoint(checkpoints=n)
    grad_mesolve(delta, adjoint, N) # jit compilation
    %timeit -n 1 -r 1 grad_mesolve(delta, adjoint, N)
```

```
DirectAdjoint
226 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)

RecursiveCheckpointAdjoint
== n = 10 ==
273 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
== n = 9 ==
261 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
== n = 8 ==
256 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
== n = 7 ==
279 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
== n = 6 ==
269 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
== n = 5 ==
294 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
== n = 4 ==
419 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
== n = 3 ==
869 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
== n = 2 ==
2.34 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
== n = 1 ==
6.75 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
```
</details>

### CODE REVIEW

**Comment by pierreguilmin** [comment]
Replaced by #459. @gautierronan I finally didn't port the "fix the dq.expect bug by using jnp.einsum", I suspect it was a bug due to the improper use of `_unvmap` in the `BacksolveAdjoint` code of diffrax.
